--- Shape.pm    Thu Mar 11 04:10:23 2004
+++ /usr/lib/perl5/site_perl/5.6.1/Geo/ShapeFile/Shape.pm       Fri Oct 15 15:48:23 2004
@@ -349,9 +349,14 @@
 
        my @p = (); # points
        while(@ps) {
-               push(@p, new Geo::ShapeFile::Point(X => shift(@ps), Y => shift(@ps)));
+#              push(@p, new Geo::ShapeFile::Point(X => shift(@ps), Y => shift(@ps)));
+                push(@p, [ shift(@ps), shift(@ps) ]);
        }
-       $self->{$label} = [@p];
+#        for (my $i = 0; $i < @ps / 2; ++$i) {
+#            $p[$i]->[0] = $ps[2 * $i];
+#            $p[$i]->[1] = $ps[2 * $i + 1];
+#        }
+       $self->{$label} = \@p;
 }
 
 sub extract_and_unpack {
